File: apps/frontend/lib/api/ats.ts
import { apiPost } from './client';

export interface ATSAnalysisResult {
  ats_score: {
    overall: number;
    breakdown: {
      keyword_match: number;
      skills_alignment: number;
      experience_relevance: number;
      education_fit: number;
      resume_completeness: number;
    };
    score_explanation: string;
  };
  keyword_analysis: {
    matched_keywords: Array<{ keyword: string; importance: string; found_in: string }>;
    missing_keywords: Array<{ keyword: string; importance: string; suggestion: string }>;
    total_jd_keywords: number;
    matched_count: number;
    match_percentage: number;
  };
  skill_gap: {
    critical_missing: Array<{ skill: string; context: string; how_to_address: string }>;
    partial_match: Array<{ skill: string; resume_has: string; jd_needs: string; gap: string }>;
    strong_matches: string[];
  };
  resume_quality: {
    completeness_score: number;
    issues: Array<{ category: string; description: string; fix: string }>;
    strengths: string[];
    bullet_quality: {
      has_action_verbs: boolean;
      has_metrics: boolean;
      average_bullet_strength: string;
      weak_bullets_count: number;
    };
  };
  interview_questions: Array<{
    question: string;
    category: string;
    why_asked: string;
    tip: string;
  }>;
  tailoring_recommendations: Array<{
    priority: string;
    section: string;
    recommendation: string;
    example: string;
  }>;
  job_fit_verdict: {
    fit_level: string;
    summary: string;
    biggest_strength: string;
    biggest_gap: string;
  };
  resume_id: string;
  job_id: string;
  analyzed_at: string;
}

export async function analyzeATSMatch(resumeId: string, jobId: string): Promise<ATSAnalysisResult> {
  const res = await apiPost('/ats/analyze', { resume_id: resumeId, job_id: jobId });
  if (!res.ok) {
    const text = await res.text().catch(() => '');
    throw new Error(`ATS analysis failed (${res.status}): ${text}`);
  }
  const payload = await res.json();
  return payload.data;
}

export interface SuggestedProject {
  name: string;
  role: string;
  years?: string;
  description: string[];
  rationale: string;
}

export async function suggestProject(resumeId: string, jobId: string): Promise<SuggestedProject> {
  const res = await apiPost('/ats/suggest-project', { resume_id: resumeId, job_id: jobId });
  if (!res.ok) {
    const text = await res.text().catch(() => '');
    throw new Error(`Failed to suggest project (${res.status}): ${text}`);
  }
  const payload = await res.json();
  return payload.project;
}

// ── Project Optimizer ─────────────────────────────────────────────────────────

export interface ProjectRelevance {
  index: number;
  name: string;
  relevance_score: number;
  verdict: 'keep' | 'replace';
  reason: string;
  jd_skills_matched: string[];
  jd_skills_missing: string[];
}

export interface AnalyzeProjectsResult {
  projects: ProjectRelevance[];
  summary: string;
}

export async function analyzeProjects(
  resumeId: string,
  jobId: string
): Promise<AnalyzeProjectsResult> {
  const res = await apiPost('/ats/analyze-projects', {
    resume_id: resumeId,
    job_id: jobId,
  });
  if (!res.ok) {
    const text = await res.text().catch(() => '');
    throw new Error(`Failed to analyze projects (${res.status}): ${text}`);
  }
  const payload = await res.json();
  return { projects: payload.projects, summary: payload.summary };
}

export interface ReplacementProject {
  id?: number;
  name: string;
  role: string;
  years?: string;
  description: string[];
  rationale: string;
}

export async function replaceProject(
  resumeId: string,
  jobId: string,
  projectIndex: number,
  replaceReason: string
): Promise<ReplacementProject> {
  const res = await apiPost('/ats/replace-project', {
    resume_id: resumeId,
    job_id: jobId,
    project_index: projectIndex,
    replace_reason: replaceReason,
  });
  if (!res.ok) {
    const text = await res.text().catch(() => '');
    throw new Error(`Failed to replace project (${res.status}): ${text}`);
  }
  const payload = await res.json();
  return payload.project;
}

File: apps/backend/app/schemas/ats.py
"""Pydantic models for ATS analysis, skill gap, and interview prep features."""

from typing import Literal
from pydantic import BaseModel, Field


# ── Keyword Analysis ──────────────────────────────────────────────────────────

class MatchedKeyword(BaseModel):
    keyword: str
    importance: Literal["critical", "important", "nice_to_have"]
    found_in: str  # skills | experience | education | summary


class MissingKeyword(BaseModel):
    keyword: str
    importance: Literal["critical", "important", "nice_to_have"]
    suggestion: str  # How to naturally add this


class KeywordAnalysis(BaseModel):
    matched_keywords: list[MatchedKeyword] = Field(default_factory=list)
    missing_keywords: list[MissingKeyword] = Field(default_factory=list)
    total_jd_keywords: int = 0
    matched_count: int = 0
    match_percentage: float = 0.0


# ── ATS Score ─────────────────────────────────────────────────────────────────

class ATSScoreBreakdown(BaseModel):
    keyword_match: int = Field(ge=0, le=100)
    skills_alignment: int = Field(ge=0, le=100)
    experience_relevance: int = Field(ge=0, le=100)
    education_fit: int = Field(ge=0, le=100)
    resume_completeness: int = Field(ge=0, le=100)


class ATSScore(BaseModel):
    overall: int = Field(ge=0, le=100)
    breakdown: ATSScoreBreakdown
    score_explanation: str


# ── Skill Gap ─────────────────────────────────────────────────────────────────

class CriticalMissingSkill(BaseModel):
    skill: str
    context: str        # Why this skill matters for this role
    how_to_address: str # Practical advice


class PartialMatchSkill(BaseModel):
    skill: str
    resume_has: str
    jd_needs: str
    gap: str


class SkillGap(BaseModel):
    critical_missing: list[CriticalMissingSkill] = Field(default_factory=list)
    partial_match: list[PartialMatchSkill] = Field(default_factory=list)
    strong_matches: list[str] = Field(default_factory=list)


# ── Resume Quality ────────────────────────────────────────────────────────────

class ResumeIssue(BaseModel):
    category: str  # missing_section | weak_bullets | no_metrics | etc.
    description: str
    fix: str


class BulletQuality(BaseModel):
    has_action_verbs: bool
    has_metrics: bool
    average_bullet_strength: Literal["weak", "moderate", "strong"]
    weak_bullets_count: int


class ResumeQuality(BaseModel):
    completeness_score: int = Field(ge=0, le=100)
    issues: list[ResumeIssue] = Field(default_factory=list)
    strengths: list[str] = Field(default_factory=list)
    bullet_quality: BulletQuality


# ── Interview Questions ───────────────────────────────────────────────────────

class InterviewQuestion(BaseModel):
    question: str
    category: Literal["technical", "behavioral", "situational", "culture_fit"]
    why_asked: str
    tip: str


# ── Tailoring Recommendations ─────────────────────────────────────────────────

class TailoringRecommendation(BaseModel):
    priority: Literal["high", "medium", "low"]
    section: str
    recommendation: str
    example: str


# ── Job Fit Verdict ───────────────────────────────────────────────────────────

class JobFitVerdict(BaseModel):
    fit_level: Literal["excellent", "good", "moderate", "poor"]
    summary: str
    biggest_strength: str
    biggest_gap: str


# ── Main Request/Response ─────────────────────────────────────────────────────

class ATSAnalysisRequest(BaseModel):
    resume_id: str
    job_id: str


class ATSAnalysisResult(BaseModel):
    """Full ATS analysis result — all computed from real resume + JD content."""
    ats_score: ATSScore
    keyword_analysis: KeywordAnalysis
    skill_gap: SkillGap
    resume_quality: ResumeQuality
    interview_questions: list[InterviewQuestion] = Field(default_factory=list)
    tailoring_recommendations: list[TailoringRecommendation] = Field(default_factory=list)
    job_fit_verdict: JobFitVerdict
    # Metadata
    resume_id: str
    job_id: str
    analyzed_at: str


class ATSAnalysisResponse(BaseModel):
    request_id: str
    data: ATSAnalysisResult
